zrml/orderbook/src/lib.rs

```diff
@@ -164,181 +178,274 @@ mod pallet {
+                    ]
+                }
+            }
+        }
+
+        /// Reduces the reserved maker and requested taker amount
+        /// by the amount the maker and taker actually filled.
+        fn decrease_order_amounts(
+            order_data: &mut OrderOf<T>,
+            maker_fill: BalanceOf<T>,
+            taker_fill: BalanceOf<T>,
+        ) -> DispatchResult {
+            order_data.maker_amount = order_data.maker_amount.checked_sub_res(&taker_fill)?;
+            order_data.taker_amount = order_data.taker_amount.checked_sub_res(&maker_fill)?;
+            Ok(())
+        }
+
+        /// Calculates the amount that the taker is going to get from the maker's amount.
+        fn get_taker_fill(
+            order_data: &OrderOf<T>,
+            maker_fill: BalanceOf<T>,
+        ) -> Result<BalanceOf<T>, DispatchError> {
+            // the maker_full_fill is the maximum amount of what the maker wants to have
+            let maker_full_fill = order_data.taker_amount;
+            // the taker_full_fill is the maximum amount of what the taker wants to have
+            let taker_full_fill = order_data.maker_amount;
+            // rounding down: the taker will always get a little bit less than what they asked for.
+            // This ensures that the reserve of the maker
+            // is always enough to repatriate successfully!
+            // `maker_full_fill` is ensured to be never zero in `ensure_ratio_quotient_valid`
+            let ratio = maker_fill.bdiv_floor(maker_full_fill)?;
+            // returns the (partial) amount of what the taker gets from the maker's amount
+            // respected the partial fill from the taker of what the maker wants to get filled
+            ratio.bmul_floor(taker_full_fill)
+        }
+
+        fn ensure_ratio_quotient_valid(order_data: &OrderOf<T>) -> DispatchResult {
+            let maker_full_fill = order_data.taker_amount;
+            // this ensures that partial fills, which fill nearly the whole order, are not executed
+            // this protects the last fill happening
+            // without a division by zero for `Perquintill::from_rational`
+            let is_ratio_quotient_valid = maker_full_fill.is_zero()
+                || maker_full_fill >= T::AssetManager::minimum_balance(order_data.taker_asset);
+            ensure!(is_ratio_quotient_valid, Error::<T>::PartialFillNearFullFillNotAllowed);
+            Ok(())
+        }
+
+        fn do_remove_order(order_id: OrderId, who: AccountIdOf<T>) -> DispatchResult {
+            let order_data = <Orders<T>>::get(order_id).ok_or(Error::<T>::OrderDoesNotExist)?;
+            debug_assert!(
+                {
+                    let market = T::MarketCommons::market(&order_data.market_id)?;
+                    market.scoring_rule == ScoringRule::Orderbook
+                },
+                "The call to place_order already ensured the scoring rule order book."
+            );
```

Comment:
Ok, I generally agree that I should reconsider my `debug_assert` usage. I just thought it would be a good way to tell the reader of the code that this assumption is true. `debug_assert` is also an excellent way to have no overhead in the production runtime, since its logic is not included in those builds. We also discussed exactly this code snippet already in the review from Malte. Before using `debug_assert`, I used a normal `ensure!` statement, but Malte suggested that we shouldn't include checks that wouldn't ever happen in production (because we ensure each order created does have the scoring rule order book). Your suggestion comes with the drawback that we would need to runtime overhead in production builds that we wouldn't necessarily need. So at the end, I am in favour of deleting this check overall. What would you do @maltekliemann ?

Reply:
> but Malte suggested that we shouldn't include checks that wouldn't ever happen in production

I think (hope) that I suggested that you be consistent in using or not using this pattern. In terms of software design, leaving the check out is preferred, in terms of defensive programming, leaving the check in is preferred. Either way is fine.

> The runtime review from SRLabs also expressed that debug_assert in production is a bad practice and should be avoided.

When/if SRLabs publishes a best practices, I'm happing to go over our style guide with you guys and incorporate these. I think there are some other things that we may want to add there.

Reply:
> I'm happing to go over our style guide with you guys and incorporate these. I think there are some other things that we may want to add there.

Happy to apply best practices and agree on that.

Reply:
537725ec679acb73e5340c7de68ff005f9d3bee3